File: modules/common-logic/src/test/java/logic/parse/tree/TokenTreeTest.java
package logic.parse.tree;

import static logic.parse.tree.CharacterClass.ALPHA_CHARACTERS;
import static logic.parse.tree.CharacterClass.DIGIT_CHARACTERS;
import static logic.parse.tree.CharacterClass.WORD_CHARACTERS;
import static logic.parse.tree.NodeExpression.DECIMAL;
import static logic.parse.tree.NodeExpression.INTEGER;
import static logic.parse.tree.NodeExpression.NUMBER;
import static logic.parse.tree.NodeExpression.WORD;
import static logic.parse.tree.NodeExpression.anyOf;
import static logic.parse.tree.NodeExpression.literal;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.List;
import logic.ResolvedValue;
import org.junit.jupiter.api.Test;

class TokenTreeTest {

    @Test
    void emptyTree() {
        assertThrows(ParseException.class, () ->
                TokenTree.<ResolvedValue>create().parse("Text"));
    }

    @Test
    void singleNode() {
        var tree = TokenTree.<ResolvedValue>create()
                .add("A", ResolvedValue::of);
        assertThat(tree.parse("A")).contains(ResolvedValue.of("A"));
    }

    @Test
    void simpleChain() {
        var tree = TokenTree.<ResolvedValue>create()
                .add("ABC", ResolvedValue::of);
        assertThat(tree.parse("ABC")).contains(ResolvedValue.of("ABC"));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse("A"));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse("AB"));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse("ABX"));
    }

    @Test
    void splittingChain() {
        var tree = TokenTree.<ResolvedValue>create()
                .add("ABC", ResolvedValue::of)
                .add("A23", ResolvedValue::of);
        assertThat(tree.parse("ABC")).contains(ResolvedValue.of("ABC"));
        assertThat(tree.parse("A23")).contains(ResolvedValue.of("A23"));
    }

    @Test
    void multipleTokens() {
        var tree = TokenTree.<ResolvedValue>create()
                .ignoreWhitespaces()
                .add("ABC", ResolvedValue::of)
                .add("123", ResolvedValue::of);
        assertThat(tree.parse("ABC 123"))
                .containsExactly(ResolvedValue.of("ABC"), ResolvedValue.of("123"));
    }

    @Test
    void anyOfToken() {
        var tree = TokenTree.<ResolvedValue>create()
                .ignoreWhitespaces()
                .add(anyOf(DIGIT_CHARACTERS), token -> ResolvedValue.of(Integer.parseInt(token)));
        assertThat(tree.parse("1 2 3"))
                .containsExactly(ResolvedValue.of(1), ResolvedValue.of(2), ResolvedValue.of(3));
    }

    @Test
    void anyOfChain() {
        var tree = TokenTree.<ResolvedValue>create()
                .ignoreWhitespaces()
                .add(List.of(anyOf(DIGIT_CHARACTERS), anyOf(DIGIT_CHARACTERS)), token -> ResolvedValue.of(Integer.parseInt(token)));
        assertThat(tree.parse("13 25 36"))
                .containsExactly(ResolvedValue.of(13), ResolvedValue.of(25), ResolvedValue.of(36));
    }

    @Test
    void repeated() {
        var tree = TokenTree.<ResolvedValue>create()
                .ignoreWhitespaces()
                .add(anyOf(DIGIT_CHARACTERS).repeats(1, 2), token -> ResolvedValue.of(Integer.parseInt(token)));
        assertThat(tree.parse("5")).containsExactly(ResolvedValue.of(5));
        assertThat(tree.parse("73")).containsExactly(ResolvedValue.of(73));
        assertThat(tree.parse("12 9 23")).containsExactly(ResolvedValue.of(12), ResolvedValue.of(9), ResolvedValue.of(23));
    }

    @Test
    void optionalTrailingCharacter() {
        var tree = TokenTree.<ResolvedValue>create()
                .ignoreWhitespaces()
                .add(List.of(anyOf(ALPHA_CHARACTERS).repeats(1), anyOf(DIGIT_CHARACTERS).optional()), ResolvedValue::of);
        assertThat(tree.parse("A5")).containsExactly(ResolvedValue.of("A5"));
        assertThat(tree.parse("ABC6")).containsExactly(ResolvedValue.of("ABC6"));
        assertThat(tree.parse("ABC")).containsExactly(ResolvedValue.of("ABC"));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse("5"));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse("ABC56"));
    }

    @Test
    void optionalLeadingCharacter() {
        var tree = TokenTree.<ResolvedValue>create()
                .ignoreWhitespaces()
                .add(List.of(anyOf("@").optional(), anyOf(WORD_CHARACTERS).repeats(1)), ResolvedValue::of);
        assertThat(tree.parse("@A5")).containsExactly(ResolvedValue.of("@A5"));
        assertThat(tree.parse("A5")).containsExactly(ResolvedValue.of("A5"));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse("@"));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse("@@A5"));
    }

    @Test
    void numberToken() {
        var tree = TokenTree.<ResolvedValue>create()
                .ignoreWhitespaces()
                .add(NUMBER, token -> ResolvedValue.of(Double.parseDouble(token)));
        assertThat(tree.parse("1")).containsExactly(ResolvedValue.of(1.0));
        assertThat(tree.parse("23")).containsExactly(ResolvedValue.of(23.0));
        assertThat(tree.parse("54890")).containsExactly(ResolvedValue.of(54890.0));
        assertThat(tree.parse("3.14")).containsExactly(ResolvedValue.of(3.14));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse("A"));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse("5B"));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse("2."));
        assertThrows(ParseException.class, () -> TokenTree.<ResolvedValue>create().parse(".5"));
    }

    @Test
    void expressionOrder() {
        var tree = TokenTree.<ResolvedValue>create()
                .ignoreWhitespaces()
                .add(INTEGER, ResolvedValue::of)
                .add(DECIMAL, token -> ResolvedValue.of(Double.parseDouble(token)));
        assertThat(tree.parse("123 3.14 5 0.2")).containsExactly(
                ResolvedValue.of("123"),
                ResolvedValue.of(3.14),
                ResolvedValue.of("5"),
                ResolvedValue.of(0.2));
    }

    @Test
    void quotedToken() {
        var tree = TokenTree.<ResolvedValue>create()
                .ignoreWhitespaces()
                .add(literal("\"", "\""), ResolvedValue::of)
                .add(WORD, ResolvedValue::of);

        assertThat(tree.parse("one two \"three four\" five")).containsExactly(
                ResolvedValue.of("one"),
                ResolvedValue.of("two"),
                ResolvedValue.of("\"three four\""),
                ResolvedValue.of("five"));
    }

    @Test
    void openCloseTagToken() {
        var tree = TokenTree.<ResolvedValue>create()
                .ignoreWhitespaces()
                .add(literal("<open>", "<close>"), ResolvedValue::of)
                .add(WORD, ResolvedValue::of);

        assertThat(tree.parse("one two <open>three four<close> five"))
                .containsExactly(
                        ResolvedValue.of("one"),
                        ResolvedValue.of("two"),
                        ResolvedValue.of("<open>three four<close>"),
                        ResolvedValue.of("five"));
    }
}

File: pathfinder-frontend/src/logic/TokenTree.test.js
import TokenTree, {
  any,
  anyof,
  decimal,
  integer,
  literal,
  number, term,
  words
} from "./TokenTree";
import {
  AlphaCharacters,
  DigitCharacters,
  WordCharacters
} from "./CharacterClasses";

test('empty tree', () => {
  const tree = new TokenTree();
  expect(() => tree.parse('Test')).toThrowError();
});

test('single node', () => {
  const tree = new TokenTree()
      .add('A', token => token);
  expect(tree.parse('A')).toEqual(['A']);
});

test('simple chain', () => {
  const tree = new TokenTree()
  .add('ABC', token => token);
  expect(tree.parse('ABC')).toEqual(['ABC']);
  expect(() => tree.parse('A')).toThrowError();
  expect(() => tree.parse('AB')).toThrowError();
  expect(() => tree.parse('ABX')).toThrowError();
});

test('splitting chain', () => {
  const tree = new TokenTree()
  .add('ABC', token => token)
  .add('A23', token => token);
  expect(tree.parse('ABC')).toEqual(['ABC']);
  expect(tree.parse('A23')).toEqual(['A23']);
});

test('multiple tokens', () => {
  const tree = new TokenTree()
  .ignoreWhitespaces()
  .add('ABC', token => token)
  .add('123', token => token);
  expect(tree.parse('ABC 123')).toEqual(['ABC', '123']);
});

test('anyof()', () => {
  const tree = new TokenTree()
  .ignoreWhitespaces()
  .add(anyof(DigitCharacters), token => parseInt(token));
  expect(tree.parse('1 2 3')).toEqual([1, 2, 3]);
  expect(() => tree.parse('A')).toThrowError();
});

test('anyof() chain', () => {
  const tree = new TokenTree()
  .ignoreWhitespaces()
  .add([ anyof(DigitCharacters), anyof(DigitCharacters) ], token => parseInt(token));
  expect(tree.parse('13 25 36')).toEqual([13, 25, 36]);
  expect(() => tree.parse('4')).toThrowError();
});

test('repeated', () => {
  const tree = new TokenTree()
  .ignoreWhitespaces()
  .add([ anyof(DigitCharacters).repeats(1, 2) ], token => parseInt(token));
  expect(tree.parse('5')).toEqual([5]);
  expect(tree.parse('73')).toEqual([73]);
  expect(tree.parse('12 9 23')).toEqual([12, 9, 23]);
});

test('optional trailing character', () => {
  const tree = new TokenTree()
  .ignoreWhitespaces()
  .add([ anyof(AlphaCharacters).repeats(1), anyof(DigitCharacters).optional() ], token => token);
  expect(tree.parse('A5')).toEqual(['A5']);
  expect(tree.parse('ABC6')).toEqual(['ABC6']);
  expect(tree.parse('ABC')).toEqual(['ABC']);
  expect(() => tree.parse('5')).toThrowError();
  expect(() => tree.parse('ABC56')).toThrowError();
});

test('optional leading character', () => {
  const tree = new TokenTree()
  .ignoreWhitespaces()
  .add([ anyof('@').optional(), anyof(WordCharacters).repeats(1) ], token => token);
  expect(tree.parse('@A5')).toEqual(['@A5']);
  expect(tree.parse('A5')).toEqual(['A5']);
  expect(() => tree.parse('@')).toThrowError();
  expect(() => tree.parse('@@A5')).toThrowError();
});

test('number', () => {
  const tree = new TokenTree()
  .ignoreWhitespaces()
  .add([ number ], token => parseFloat(token));
  expect(tree.parse('1')).toEqual([1]);
  expect(tree.parse('23')).toEqual([23]);
  expect(tree.parse('54890')).toEqual([54890]);
  expect(tree.parse('3.14')).toEqual([3.14]);
  expect(() => tree.parse('A')).toThrowError();
  expect(() => tree.parse('5B')).toThrowError();

  expect(() => tree.parse('2.')).toThrowError();
  expect(() => tree.parse('.5')).toThrowError();
});

test('expression order', () => {
  const tree = new TokenTree()
  .ignoreWhitespaces()
  .add(integer, token => token)
  .add(decimal, token => parseFloat(token));
  expect(tree.parse('123 3.14 5 0.2')).toEqual(['123', 3.14, '5', 0.2]);
});

test('quoted token', () => {
  const tree = new TokenTree()
  .ignoreWhitespaces()
  .add(literal('"', '"'), quote => quote)
  .add(words, word => word);

  expect(tree.parse('one two "three four" five')).toEqual(['one', 'two', '"three four"', 'five']);
});

test('open/close tag token', () => {
  const tree = new TokenTree()
  .ignoreWhitespaces()
  .add(literal('<open>', '<close>'), quote => quote)
  .add(words, word => word);

  expect(tree.parse('one two <open>three four<close> five')).toEqual(['one', 'two', '<open>three four<close>', 'five']);
});


test('+1 keen vs +2', () => {
  testAttack(25, 2 - 3, 2 + 9, 18);
  testAttack(25, 1 - 3, 1 + 9 + 3.5, 18);
});

function testAttack(targetAC, attackBonus, damageBonus, critRange) {
  const strengthMod = 6;
  const bab = 8;

  const critMultiplier = 2;

  const attackMod = strengthMod + bab;
  const damageMod = strengthMod * 1.5;

  const attempts = 100000;
  let attacks = 0;
  let rounds = 0;
  let hitCount = 0.0;
  let damageSum = 0.0;
  let critCount = 0;
  for (let i = 0; i < attempts; i++) {
    const attackResult = attack(targetAC,
        attackBonus + attackMod,
        damageBonus + damageMod,
        critRange,
        critMultiplier);
    hitCount += attackResult.hit ? 1.0 : 0.0;
    damageSum += attackResult.damage;
    critCount += attackResult.crit ? 1 : 0;
    attacks++;

    const secondAttackResult = attack(targetAC,
        attackBonus + attackMod - 5,
        damageBonus + damageMod,
        critRange,
        critMultiplier);
    hitCount += secondAttackResult.hit ? 1.0 : 0.0;
    damageSum += secondAttackResult.damage;
    critCount += secondAttackResult.crit ? 1 : 0;
    attacks++;

    rounds++;
  }

  const averageHit = hitCount / attacks;
  const averageCrits = critCount / attacks;
  const averageDamagePerHit = damageSum / hitCount;
  const totalAverageDamage = damageSum / rounds;

  console.log(`Average Hits: ${averageHit * 100}%\nAverage Crits: ${averageCrits * 100}%\nAverage Damage per Hit: ${averageDamagePerHit}\nTotal average damage: ${totalAverageDamage}`)
}

function roll(dice, sides) {
  let total = 0;
  for (let i = 0; i < dice; i++) {
    total += Math.floor(Math.random() * sides) + 1;
  }
  return total;
}

function attack(targetAC, attackBonus, damageBonus, critRange, critMultiplier) {
  const attackRoll = roll(1, 20);
  const confirmRoll = roll(1, 20);

  const attackAmount = attackRoll + attackBonus;
  const confirmAmount = confirmRoll + attackBonus;

  const isHit = (attackRoll === 20 || attackAmount >= targetAC) && attackRoll !== 1;
  const isCrit = isHit && attackRoll >= critRange;
  const isCritConfirm = isCrit && confirmRoll !== 1 && (confirmAmount >= targetAC || confirmRoll === 20);

  const damageMultiplier = isHit
      ? isCritConfirm
            ? critMultiplier
            : 1
      : 0;

  let damageAmount = 0;

  for (let i = 0; i < damageMultiplier; i++) {
    damageAmount += roll(2, 4) + damageBonus;
  }

  return {
    hit: isHit,
    damage: damageAmount,
    crit: isCritConfirm,
  };
}